proof/refine/KHeap_R.thy

```diff
@@ -11,4 +11,2094 @@ imports
+
+lemma getObject_ko_at_tcb [wp]:
+  "\<lbrace>\<top>\<rbrace> getObject p \<lbrace>\<lambda>rv::tcb. ko_at' rv p\<rbrace>"
+  by (rule getObject_ko_at | simp add: gen_objBits_simps)+
+
+lemma OMG_getObject_tcb:
+  "\<lbrace>obj_at' P t\<rbrace> getObject t \<lbrace>\<lambda>(tcb :: tcb) s. P tcb\<rbrace>"
+  apply (rule obj_at_getObject)
+  apply (clarsimp simp: loadObject_default_def in_monad)
+  done
+
+lemma setObject_nosch:
+  assumes x: "\<And>p q n ko. \<lbrace>\<lambda>s. P (ksSchedulerAction s)\<rbrace> updateObject val p q n ko \<lbrace>\<lambda>rv s. P (ksSchedulerAction s)\<rbrace>"
+  shows      "\<lbrace>\<lambda>s. P (ksSchedulerAction s)\<rbrace> setObject t val \<lbrace>\<lambda>rv s. P (ksSchedulerAction s)\<rbrace>"
+  apply (simp add: setObject_def split_def)
+  apply (wp x | simp)+
+  done
+
+lemma getObject_ep_inv:
+  "\<lbrace>P\<rbrace> (getObject addr :: endpoint kernel) \<lbrace>\<lambda>rv. P\<rbrace>"
+  by (rule getObject_inv, simp add: loadObject_default_inv)
+
+lemma getObject_ntfn_inv:
+  "\<lbrace>P\<rbrace> (getObject addr :: Structures_H.notification kernel) \<lbrace>\<lambda>rv. P\<rbrace>"
+  by (rule getObject_inv, simp add: loadObject_default_inv)
+
+lemma get_ep_inv'[wp]: "\<lbrace>P\<rbrace> getEndpoint ep \<lbrace>\<lambda>rv. P\<rbrace>"
+  by (simp add: getEndpoint_def getObject_ep_inv)
+
+lemma get_ntfn_inv'[wp]: "\<lbrace>P\<rbrace> getNotification ntfn \<lbrace>\<lambda>rv. P\<rbrace>"
+  by (simp add: getNotification_def getObject_ntfn_inv)
+
+lemma get_ep'_valid_ep[wp]:
+  "\<lbrace> invs' and ep_at' ep \<rbrace> getEndpoint ep \<lbrace> valid_ep' \<rbrace>"
+  apply (simp add: getEndpoint_def)
+  apply (rule hoare_chain)
+  apply (rule getObject_valid_obj)
+     apply simp
+    apply (simp add: gen_objBits_simps)
+   apply clarsimp
+  apply (simp add: valid_obj'_def)
+  done
+
+lemma get_ntfn'_valid_ntfn[wp]:
+  "\<lbrace> invs' and ntfn_at' ep \<rbrace> getNotification ep \<lbrace> valid_ntfn' \<rbrace>"
+  apply (simp add: getNotification_def)
+  apply (rule hoare_chain)
+  apply (rule getObject_valid_obj)
+     apply simp
+    apply (simp add: gen_objBits_simps)
+   apply clarsimp
+  apply (simp add: valid_obj'_def)
+  done
+
+lemma setEndpoint_pred_tcb_at'[wp]:
+  "\<lbrace>pred_tcb_at' proj P t\<rbrace> setEndpoint ptr val \<lbrace>\<lambda>rv. pred_tcb_at' proj P t\<rbrace>"
+  apply (simp add: pred_tcb_at'_def setEndpoint_def)
+  apply (rule obj_at_setObject2)
+  apply (clarsimp simp: updateObject_default_def in_monad)
+  done
+
+lemma get_ntfn_ko':
+  "\<lbrace>\<top>\<rbrace> getNotification ep \<lbrace>\<lambda>rv. ko_at' rv ep\<rbrace>"
+  apply (simp add: getNotification_def)
+  apply (rule getObject_ko_at)
+   apply simp
+  apply (simp add: gen_objBits_simps)
+  done
+
+lemmas ps_clear_def3 = ps_clear_def2[OF order_less_imp_le[OF aligned_less_plus_1]]
+
+declare diff_neg_mask[simp del] (* FIXME arch-split: do we want to keep this? *)
```

Comment:
Is this something that we can decide now that KHeap_R is finished?

Reply:
Yeah, with that being the lemma my preference would be to remove this but then I'm not the one who would have to do the update cycle. Maybe if you remember you could try removing it when you're having to update things anyway for something else?

Reply:
done, still waiting on CRefine